src/kernel/include/pgtbl.h

```diff
@@ -57,245 +70,398 @@ struct tlb_quiescence {
- * use physical addresses; when doing copy, we don't need to worry
- * about PA. */
-
-/* v should include the desired flags */
-static inline int
-__pgtbl_setleaf(struct ert_intern *a, void *v)
-{
-	u32_t new, old;
-
-	old = (u32_t)(a->next);
-	new = (u32_t)(v);
-
-	if (!cos_cas((unsigned long *)a, old, new)) return -ECASFAIL;
-
-	return 0;
-}
-
-/* This takes an input parameter as the old value of the mapping. Only
- * update when the existing value matches. */
-static inline int
-__pgtbl_update_leaf(struct ert_intern *a, void *v, u32_t old)
-{
-	u32_t new;
-
-	new = (u32_t)(v);
-	if (!cos_cas((unsigned long *)a, old, new)) return -ECASFAIL;
-
-	return 0;
-}
-
-/* Note:  We're just using pre-defined default flags for internal (pgd) entries */
-static int
-__pgtbl_set(struct ert_intern *a, void *v, void *accum, int isleaf)
-{
-	u32_t old, new;
-	(void)accum; assert(!isleaf);
-
-	old = (u32_t)a->next;
-	new = (u32_t)chal_va2pa((void*)((u32_t)v & PGTBL_FRAME_MASK)) | PGTBL_INTERN_DEF;
-
-	if (!cos_cas((unsigned long *)&a->next, old, new)) return -ECASFAIL;
-
-	return 0;
-}
-
-static inline void *__pgtbl_getleaf(struct ert_intern *a, void *accum)
-{ if (unlikely(!a)) return NULL; return __pgtbl_get(a, accum, 1); }
-
 ERT_CREATE(__pgtbl, pgtbl, PGTBL_DEPTH, PGTBL_ORD, sizeof(int*), PGTBL_ORD, sizeof(int*), NULL, \
 	   __pgtbl_init, __pgtbl_get, __pgtbl_isnull, __pgtbl_set,	\
 	   __pgtbl_a, __pgtbl_setleaf, __pgtbl_getleaf, ert_defresolve);
+*/
+/* How many bits have been translated ? */
+/*
+first layer, size_order 2^32, 5 bits to represent. However we only allow 8-pgtbls, thus, the last layer is fixed to 8.
+
+entry_order? ignore, because entry_order is really sucky.
+If the type is MPUMETA, then the type_addr will be the address of the next pgtbl structure.
+what about the refcnt? add refcnt when being referenced?
+we do not check refcnts? no.
+start addr, size order, entry_order. minimal page is 256B or etc. 2<<16, even 8
+size - minimal is 128 byte pages. -
+128 256 512 1k 2k 4k 8k 16k 32k 64k 128k 256k 512k 1M 2M 4M
+ 0   1   2  3  4  5  6   7   8   9   10   11   12  13 14 15
+
+ 2   4   8  16 32 64 128 256
+ 0   1   2  3  4  5   6   7
+
+The size of a page can be 2^16=
+start addr[31:8] size_order[7:4] num_order[3:1] Type[0]
+For leaf nodes:
+start_addr[31:8] status[7:0]
+For internal nodes:
+ptr [31:2] flag[1:0]
+*/
+#define COS_PGTBL_MPUMETA          0x00
+#define COS_PGTBL_TABLE            0x01
+
+#define COS_PGTBL_METAADDR(X)      ((X)&(~(0x03)))
+
+#define COS_PGTBL_STARTADDR(X)     ((X)&(0xFFFFFF00))
+#define COS_PGTBL_SIZEORDER(X)     ((((X)&0xFF)>>3)+7)
+#define COS_PGTBL_NUMORDER(X)      ((((X)&0x0E)>>1)+1)
+#define COS_PGTBL_TYPE(X)          ((X)&0x01)
+
+#define COS_PGTBL_TYPEADDR(ADDR,SIZE_ORDER,NUM_ORDER,TYPE)    (((ADDR)&(~0xFF))|((SIZE_ORDER-7)<<3)|((NUM_ORDER-1)<<1)|(TYPE))
+#define COS_PGTBL_CHILD(ADDR,FLAGS)                           ((((u32_t)(ADDR))&(~0x03))|(FLAGS))
```

Comment:
Are these macros *creating* the bits for a child or a leaf?  If so, they are named oddly.  I would expect a verb in the name to denote the action.  `COS_PGTBL_CHILD_INIT` or something?  Applies to macros around this as well.